smart_open/gcs.py

```diff
@@ -0,0 +1,523 @@
+
+    def __repr__(self):
+        return (
+            "%s("
+            "bucket=%r, "
+            "blob=%r, "
+            "buffer_size=%r)"
+        ) % (
+            self.__class__.__name__,
+            self._bucket.name,
+            self._blob.name,
+            self._buffer_size,
+        )
+
+
+class BufferedOutputBase(io.BufferedIOBase):
+    """Writes bytes to GCS.
+
+    Implements the io.BufferedIOBase interface of the standard library."""
+
+    def __init__(
+            self,
+            bucket,
+            blob,
+            min_part_size=_DEFAULT_MIN_PART_SIZE,
+            client=None,  # type: google.cloud.storage.Client
+    ):
+        if client is None:
+            client = google.cloud.storage.Client()
+        self._client = client
+        self._credentials = self._client._credentials
+        self._bucket = self._client.bucket(bucket)  # type: google.cloud.storage.Bucket
+        self._blob = self._bucket.blob(blob)  # type: google.cloud.storage.Blob
+        assert min_part_size % _MIN_MIN_PART_SIZE == 0, 'min part size must be a multiple of 256KB'
+        self._min_part_size = min_part_size
+
+        self._total_size = 0
+        self._total_parts = 0
+        self._buf = io.BytesIO()
+
+        self._session = google_requests.AuthorizedSession(self._credentials)
+
+        #
+        # https://cloud.google.com/storage/docs/json_api/v1/how-tos/resumable-upload#start-resumable
+        #
+        self._resumeable_upload_url = self._blob.create_resumable_upload_session()
+
+        #
+        # This member is part of the io.BufferedIOBase interface.
+        #
+        self.raw = None
+
+    def flush(self):
+        pass
+
+    #
+    # Override some methods from io.IOBase.
+    #
+    def close(self):
+        logger.debug("closing")
+        if self._total_size == 0:  # empty files
+            self._upload_empty_part()
+        if self._buf.tell():
+            self._upload_next_part()
+        logger.debug("successfully closed")
+
+    def writable(self):
+        """Return True if the stream supports writing."""
+        return True
+
+    def tell(self):
+        """Return the current stream position."""
+        return self._total_size
+
+    #
+    # io.BufferedIOBase methods.
+    #
+    def detach(self):
+        raise io.UnsupportedOperation("detach() not supported")
+
+    def write(self, b):
+        """Write the given bytes (binary string) to the GCS file.
+
+        There's buffering happening under the covers, so this may not actually
+        do any HTTP transfer right away."""
+
+        if not isinstance(b, _BINARY_TYPES):
+            raise TypeError(
+                "input must be one of %r, got: %r" % (_BINARY_TYPES, type(b)))
+
+        self._buf.write(b)
+        self._total_size += len(b)
+
+        if self._buf.tell() >= self._min_part_size:
+            self._upload_next_part()
+
+        return len(b)
+
+    def terminate(self):
```

Comment:
This was something I copied from the s3 module. I think having some way to abort an upload seems valuable, although it would probably be used very infrequently. I will document it